packages/source-iotsitewise/src/time-series-data/fetchTimeSeriesData.ts

```diff
@@ -0,0 +1,32 @@
+import { TimeSeriesDataModule, Viewport, DataStream } from '@iot-app-kit/core';
+import { SiteWiseDataStreamQuery } from './types';
+
+const MAX_RETRY_TIME = 9000;
+
+const waitForRetry = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+
+// creates async function which requests cachedDataStreams until all the data for viewport is present
+export const fetchTimeSeriesData =
+  (siteWiseTimeSeriesModule: TimeSeriesDataModule<SiteWiseDataStreamQuery>) =>
+  async ({ query, viewport }: { query: SiteWiseDataStreamQuery; viewport: Viewport }): Promise<DataStream[]> => {
+    // failsafe to break out of loop if data isnt present after 9 seconds
+    const firstRetryTime = Date.now();
+
+    // run loops on 500ms interval until all data in range is found OR 9 seconds pass
+    for (; ; await waitForRetry(500)) {
+      const { dataStreams, hasAllDataInRange } = await siteWiseTimeSeriesModule.getCachedDataStreams({
```

Comment:
I think a better approach would be to somehow subscribe to the redux data store, rather than polling it. Check out over here, with observable store: https://github.com/awslabs/iot-app-kit/blob/main/packages/core/src/data-module/data-cache/dataCacheWrapped.ts#L76